[FEATURE] IPC4: verify-sof-firmware-load.sh fails

This test is unable to find out if the firmware was successfully loaded with hard-coded dependencies on 'sof

````starting test_firmware-load 
+ test_firmware-load
+ /root/sof-test/test-case/verify-sof-firmware-load.sh
2022-01-27 23:55:28 UTC Sub-Test: [INFO] Checking SOF Firmware load info in kernel log
[   61.608284] kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: ffffff10 00000001 00000000 00001002
[   61.608287] kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: 00000000 00000000 00000180 00000001
[   61.608289] kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: 00000000
[   61.608292] kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: ipc4 create module 6 - instance 0 - pipe 1 - core 0 - domain 0 - param size 64
[   61.608788] kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: 00000000 00000180 00000180 00000000
[   61.608795] kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: 0000bb80 00000010 ffffff10 00000001
[   61.608799] kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: 00000000 00001002 ffffffff 7fffffff
[   61.608802] kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: 00000000 00000000 0000000a 00000000
[   61.608806] kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: ipc4 create module 4 - instance 1 - pipe 1 - core 0 - domain 0 - param size 84
[   61.609439] kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: 0000066f 000000c0 000000c0 00000001
[   61.609446] kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: 0000bb80 00000010 ffffff10 00000001
[   61.609450] kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: 00000000 00001002 0000bb80 00000010
[   61.609453] kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: ffffff10 00000001 00000000 00001002
[   61.609457] kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: 00000000 00000800 00000180 00000001
[   61.609459] kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: 00000000
[   61.609463] kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: ipc4 bind module 4 -0 to module 6 -0
[   61.609795] kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: route copier.host.1.1 -> gain.1.1 setup complete
[   61.609800] kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: ipc4 bind module 6 -0 to module 4 -1
[   61.610112] kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: route gain.1.1 -> copier.HDA.1.0 setup complete
[   61.610116] kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: pcm: stream dir 0, posn mailbox offset is 0
[   61.611151] kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: In hda_link_pcm_trigger cmd=1
[   61.611167] kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: pcm: trigger stream 0 dir 0 cmd 1
[   61.611180] kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: FW Poll Status: reg[0x160]=0x14001e successful
[   61.611187] kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: ipc4 set pipeline 1 status 3
[   61.611691] kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: ipc4 set pipeline 1 status 4
[   61.949031] kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: pcm: trigger stream 0 dir 0 cmd 0
[   61.949042] kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: ipc4 set pipeline 1 status 3
[   61.949825] kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: ipc4 set pipeline 1 status 2
[   61.950181] kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: FW Poll Status: reg[0x160]=0x140000 successful
[   61.950213] kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: In hda_link_pcm_trigger cmd=0
[   61.950281] kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: pcm: free stream 0 dir 0
[   61.950304] kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: ipc4 unbind module 4 -0 to module 6 -0
[   61.950615] kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: route copier.host.1.1 -> gain.1.1 unbind complete
[   61.950624] kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: widget copier.host.1.1 instance 0 freed
[   61.950631] kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: ipc4 unbind module 6 -0 to module 4 -1
[   61.951011] kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: route gain.1.1 -> copier.HDA.1.0 unbind complete
[   61.951019] kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: widget gain.1.1 instance 0 freed
[   61.951026] kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: widget copier.HDA.1.0 instance 1 freed
[   61.951032] kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: ipc4 delete pipeline 1
[   61.951866] kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: widget pipeline.1 instance 0 freed
[   61.952010] kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: pcm: close stream 0 dir 0
[   64.247779] kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: FW Poll Status: reg[0x4]=0x1010f0f successful
[   64.250480] kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: FW Poll Status: reg[0x4]=0xf0f successful
[   64.250486] kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: DSP core(s) enabled? 0 : core_mask 1
[   64.250559] kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: Debug PCIR: 00000010 at  00000044
[   64.250565] kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: Turning i915 HDAC power 0
[   64.250567] kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: Current DSP power state: D3
[   64.250568] kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: fw_state change: 6 -> 0
[44746.949373] kernel: perf: interrupt took too long (2501 > 2500), lowering kernel.perf_event_max_sample_rate to 79000
[77952.212622] kernel: perf: interrupt took too long (3127 > 3126), lowering kernel.perf_event_max_sample_rate to 63000
 ------
  Check journalctl status: 
 ---- 
● systemd-journald.service - Journal Service
     Loaded: loaded (/usr/lib/systemd/system/systemd-journald.service; static)
     Active: active (running) since Wed 2022-01-26 15:18:06 PST; 24h ago
TriggeredBy: ● systemd-journald.socket
             ● systemd-journald-audit.socket
             ● systemd-journald-dev-log.socket
       Docs: man:systemd-journald.service(8)
             man:journald.conf(5)
   Main PID: 468 (systemd-journal)
     Status: "Processing requests..."
        CPU: 895ms
     CGroup: /system.slice/systemd-journald.service
             └─468 /usr/lib/systemd/systemd-journald

Jan 26 15:18:06 fedora systemd-journald[468]: Journal started
Jan 26 15:18:06 fedora systemd-journald[468]: Runtime Journal (/run/log/journal/aeb6d152cc6b470ea11ccfb36c4d966c) is 8.0M, max 1.2G, 1.2G free.
Jan 26 15:18:06 fedora systemd[1]: systemd-journald.service: unit configures an IP firewall, but the local system does not support BPF/cgroup firewalling.
Jan 26 15:18:06 fedora systemd[1]: (This warning is only shown for the first unit using IP firewalling.)
Jan 26 15:18:06 fedora systemd[1]: systemd-journald.service: Deactivated successfully.
Jan 26 15:18:06 fedora systemd-journald[468]: Time spent on flushing to /var/log/journal/aeb6d152cc6b470ea11ccfb36c4d966c is 21.055ms for 1147 entries.
Jan 26 15:18:06 fedora systemd-journald[468]: System Journal (/var/log/journal/aeb6d152cc6b470ea11ccfb36c4d966c) is 1.0G, max 2.8G, 1.7G free.

● systemd-journald-audit.socket - Journal Audit Socket
     Loaded: loaded (/usr/lib/systemd/system/systemd-journald-audit.socket; static)
     Active: active (running) since Wed 2022-01-26 07:18:03 PST; 1 day 8h ago
   Triggers: ● systemd-journald.service
       Docs: man:systemd-journald.service(8)
             man:journald.conf(5)
     Listen: audit 1 (Netlink)
     CGroup: /system.slice/systemd-journald-audit.socket

Notice: journal has been rotated since unit was started, output may be incomplete.

● systemd-journald.socket - Journal Socket
     Loaded: loaded (/usr/lib/systemd/system/systemd-journald.socket; static)
     Active: active (running) since Wed 2022-01-26 07:18:03 PST; 1 day 8h ago
   Triggers: ● systemd-journald.service
       Docs: man:systemd-journald.service(8)
             man:journald.conf(5)
     Listen: /run/systemd/journal/socket (Datagram)
             /run/systemd/journal/stdout (Stream)
     CGroup: /system.slice/systemd-journald.socket

Notice: journal has been rotated since unit was started, output may be incomplete.

● systemd-journald-dev-log.socket - Journal Socket (/dev/log)
     Loaded: loaded (/usr/lib/systemd/system/systemd-journald-dev-log.socket; static)
     Active: active (running) since Wed 2022-01-26 07:18:03 PST; 1 day 8h ago
   Triggers: ● systemd-journald.service
       Docs: man:systemd-journald.service(8)
             man:journald.conf(5)
     Listen: /run/systemd/journal/dev-log (Datagram)
     CGroup: /system.slice/systemd-journald-dev-log.socket

Notice: journal has been rotated since unit was started, output may be incomplete.
2022-01-27 23:55:29 UTC Sub-Test: [ERROR] Cannot find the sof audio version
2022-01-27 23:55:29 UTC Sub-Test: [ERROR] Starting func_exit_handler(), exit status=1, FUNCNAME stack:
2022-01-27 23:55:29 UTC Sub-Test: [ERROR]  die()  @  /root/sof-test/test-case/../case-lib/lib.sh
2022-01-27 23:55:29 UTC Sub-Test: [ERROR]  main()  @  /root/sof-test/test-case/verify-sof-firmware-load.sh:40
2022-01-27 23:55:29 UTC Sub-Test: [INFO] Test Result: FAIL!
^C

Durations in seconds:
	 durations=([firmware-load]="1" [firmware-presence]="0" )


PASS: firmware-presence;
FAIL: firmware-load;

````

--- a/case-lib/lib.sh
+++ b/case-lib/lib.sh
@@ -443,6 +443,17 @@ print_module_params()
     echo "----------------------------------------"
 }
 
+# "$@" is optional, typically: --since=@"$epoch".
+# shellcheck disable=SC2120
+grep_firmware_info_in_logs()
+{
+    # dump the version info and ABI info
+    # "head -n" makes this compatible with set -e.
+    journalctl_cmd "$@" | grep "Firmware info" -A1 | head -n 12
+    # dump the debug info
+    journalctl_cmd "$@" | grep "Firmware debug build" -A3 | head -n 12
+}
+
 # check if NTP Synchronized, if so return 0 otherwise return 1
 check_ntp_sync()
 {

--- a/case-lib/lib.sh
+++ b/case-lib/lib.sh
@@ -395,6 +395,13 @@ disable_kernel_check_point()
     KERNEL_CHECKPOINT="disabled"
 }
 
+# "$@" is optional, usually: --since=@"$epoch_checkpoint"
+# shellcheck disable=SC2120
+sof_firmware_boot_complete()
+{
+    journalctl_cmd "$@" --grep 'sof.*firmware[[:blank:]]*boot[[:blank:]]*complete'
+}
+
 is_zephyr()
 {
     local ldcFile

--- a/test-case/check-kmod-load-unload.sh
+++ b/test-case/check-kmod-load-unload.sh
@@ -91,8 +91,11 @@ do
         die "Found error(s) in kernel log after module insertion"
 
     dlogi "checking if firmware is loaded successfully"
-    "$(dirname "${BASH_SOURCE[0]}")"/verify-sof-firmware-load.sh ||
+    if sof_firmware_boot_complete --since=@"$KERNEL_CHECKPOINT"; then
+        grep_firmware_info_in_logs --since=@"$KERNEL_CHECKPOINT"
+    else
          die "Failed to load firmware after module insertion"
+    fi
 
     # successful remove/insert module pass
     dlogi "==== firmware boot complete: $idx of $loop_cnt ===="

--- a/test-case/verify-sof-firmware-load.sh
+++ b/test-case/verify-sof-firmware-load.sh
@@ -26,10 +26,8 @@ cmd="journalctl_cmd"
 
 dlogi "Checking SOF Firmware load info in kernel log"
 if $cmd | grep -q " sof-audio.*Firmware.*version"; then
-    # dump the version info and ABI info
-    $cmd | grep "Firmware info" -A1 | head -n 12
-    # dump the debug info
-    $cmd | grep "Firmware debug build" -A3 | head -n 12
+
+    grep_firmware_info_in_logs
     exit 0
 
 else # failed, show some logs

--- a/test-case/verify-sof-firmware-load.sh
+++ b/test-case/verify-sof-firmware-load.sh
@@ -25,7 +25,7 @@ setup_kernel_check_point
 cmd="journalctl_cmd"
 
 dlogi "Checking SOF Firmware load info in kernel log"
-if $cmd | grep -q " sof-audio.*Firmware.*version"; then
+if sof_firmware_boot_complete; then
 
     grep_firmware_info_in_logs
     exit 0

